Updates to cert manager to break bolt db dependency

aalvarado1 opened the pull request:
The cert manager will now use an external PostgreSQL DB to manage persistent data. New command line arg has been added to run service.

./kube-cert-manager -default-email=some@email.com -kubeconfig=/Users/user1/.kube/config -acme-url=http://localhost:4000/directory -data-dir=/Users/folder1/data -db-args="host=localhost port=5432 user=user1 dbname=certmanagerdb password=password sslmode=disable"'

Commit:
Initial start of refactor to use postgreSQL instead of boltDB.

Commit:
Updates to the db method calls. Still work in progress. Still need to
add logging, error handling, and clean up

Commit:
Updates to database query functions to return proper byte arrays. Investigating an error when returning alt-names byte array.

Commit:
If alt names is empty db return an empty byte array. Update to check for empty alt-name byte array

Commit:
Code clean up, remove print and commented out code.

Commit:
Addition of update postgres calls for when a cert needs to be renewed.
Also some updates for nil check on the byte arrays causing issues.

Commit:
Added config params for configuration of external database.

Commit:
Clean up and close connection to db one the app has signal to close.

ctriv commented on db.go:
```diff
@@ -0,0 +1,133 @@
+	log.Printf("Retreving domain alt-names from database for domain (%s)", domain)
+
+	var altNames DomainAltname
+	d.Where(&DomainAltname{Domain: domain}).First(&altNames)
```
You should check for DB errors here
```suggestion
	err = d.Where(&DomainAltname{Domain: domain}).First(&altNames).Error
	if err != nil {
	    return nil, err
	}
```

aalvarado1 replied:
Question about the err check. Is it not okay to pass the error up to the processor and have it log it?
`userInfoRaw, err = getUserInfo(email, p.db)

	if err != nil {
		return p.NoteCertError(cert, err, "Error while running database view user information transaction for domain %v", cert.Spec.Domain)
	}`
In this case the getUserInfo is called which could contain an error but it will be part of the p.NoteCertError?

ctriv replied:
There's nothing wrong with passing the error up, that's a typical way of handling errors in go.  If you need to add some extra context information to the error, `errors.Wrap` or `errors.Wrapf` is often used.  My point was that the code wasn't checking gorm for errors, it was always returning nil for the error.

aalvarado1 replied:
Sounds good. I am going to wrap the gorm error into the reg err.

ctriv commented on db.go:
```diff
@@ -0,0 +1,133 @@
+Get Domain AltNames via domain
+*/
+func getAltNames(domain string, d *gorm.DB) (altNamesRaw []byte, err error) {
+	log.Printf("Retreving domain alt-names from database for domain (%s)", domain)
```
This will likely get pretty spamy in production, I'd think about taking it out.

aalvarado1 replied:
Removing spam, see above comment about error handling.

ctriv commented on db.go:
```diff
@@ -0,0 +1,133 @@
+/**
+Get User Info via email
+*/
+func getUserInfo(email string, d *gorm.DB) (userInfoRaw []byte, err error) {
```
Spam and error handling in this function too.

aalvarado1 replied:
Removing spam, see above comment about error handling.

ctriv commented on db.go:
```diff
@@ -0,0 +1,133 @@
+Get Certificate Details via Domain
+*/
+func getCertDetails(domain string, d *gorm.DB) (certDetailsRaw []byte, err error) {
+	log.Printf("Retreving certificate details from database for domain (%s)", domain)
```
Spam and errors.

aalvarado1 replied:
Removing spam, see above comment about error handling.

ctriv commented on db.go:
```diff
@@ -0,0 +1,133 @@
+Save User Information key email, value userInfo
+*/
+func addUserInfo(email string, userInfo []byte, d *gorm.DB) (err error) {
+	log.Printf("Saving user info email (%s) to database", email)
```
Spam and errors.

aalvarado1 replied:
Removing spam, see above comment about error handling.

ctriv commented on db.go:
```diff
@@ -0,0 +1,133 @@
+Save Certificate Details key domain, value certDetails
+*/
+func addCertDetails(domain string, certDetails []byte, d *gorm.DB) (err error) {
+	log.Printf("Saving certificate details to database for domain (%s)", domain)
```
Spam and errors.

aalvarado1 replied:
Removing spam, see above comment about error handling.

ctriv commented on db.go:
```diff
@@ -0,0 +1,133 @@
+Save Alt Names Details key domain, value altNames
+*/
+func addAltNames(domain string, altNames []byte, d *gorm.DB) (err error) {
+	log.Printf("Saving domain alt-names to database for domain (%s)", domain)
```
Spam and errors.

aalvarado1 replied:
Removing spam, see above comment about error handling.

ctriv commented on db.go:
```diff
@@ -0,0 +1,133 @@
+Update Certificate Details key domain, new value certDetails
+*/
+func updateCertDetails(domain string, certDetailsRaw []byte, d *gorm.DB) (err error) {
+	log.Printf("Updating domain alt-names to database for domain (%s)", domain)
```
Spam and errors.

aalvarado1 replied:
Removing spam, see above comment about error handling.

ctriv commented on main.go:
```diff
@@ -81,40 +78,18 @@ func main() {
 	flag.IntVar(&renewBeforeDays, "renew-before-days", 7, "Renew certificates before this number of days until expiry")
 	flag.IntVar(&workers, "workers", 4, "Number of parallel jobs to run at once")
+	flag.StringVar(&dbArgs, "db-args", "host=localhost port=5432 user=certmanager dbname=certmanager password=Pass1234 sslmode=disable",
+		"db host, port, username, password, sslmode")
```
This feels a little awkward. Would it be be better to have `db-host`, `db-user`, etc..?  Down the road we should think about moving to a configuration system like [viper](https://github.com/spf13/viper), but that's likely outside the scope of this effort.

aalvarado1 replied:
okay sound good, I'll break that up.

ctriv commented on main.go:
```diff
@@ -81,40 +78,18 @@ func main() {
 	flag.StringVar(&defaultEmail, "default-email", "", "Default email address for ACME registrations")
```
You should no longer need the `data-dir` flag.

aalvarado1 replied:
Ah yes, good catch no boltdb no need for data-dir

ctriv commented on main.go:
```diff
@@ -196,6 +173,7 @@ func main() {
 	signal.Notify(signalChan, syscall.SIGINT, syscall.SIGTERM)
 	<-signalChan
 	log.Println("Shutdown signal received, exiting...")
+	p.db.Close()
```
Consider using a `defer` for this right after you open the db connection.

aalvarado1 replied:
If I add a defer right after the db connection that connection will be close and no operation will be able to be preformed. I can either open and close the connection for each call or I can keep the connection open as advised https://github.com/jinzhu/gorm/issues/1427. This placement will keep the connection open until the shutdown signal is received.

ctriv replied:
defer does not execute until main() returns.  It's the idiomatic way in go to deal with cleaning up resources.  See https://tour.golang.org/flowcontrol/12

aalvarado1 replied:
Yes that is true but because I broke out the db operation into functions and pass around the pointer to the db, I think main thinks there is nothing to do so it closes the connection. I tried a tests and it does close the connection.
`(sql: database is closed) 
[2018-11-09 14:28:56]  
`

aalvarado1 replied:
Maybe I'm missing something... I have it in the db method... maybe I need to create a main function and put the open/close there?

ctriv replied:
Defered calls are only executed when the enclosing function is done, if the defer on the db close from main isn't working then something funky is going on.  I can take a look next week, but it's worrisome that the defer isn't working.

aalvarado1 replied:
Well it does work if I put all operation in the same function or in main. I'll mess around with it. We can have a hangout on Monday and go over it.

ctriv commented on db.go:
```diff
@@ -0,0 +1,133 @@
+	}
+
+	return db, err
+}
```
You'll never return an error from this function, if err isn't nill, you panic.  You should remove the returned error or the panic.

aalvarado1 replied:
Yes, that make sense.

russellmacshane commented on processor.go:
```diff
@@ -624,15 +623,26 @@ func (p *CertProcessor) processCertificate(cert Certificate, forMaint bool) (boo
+		err = updateCertDetails(cert.Spec.Domain, certDetailsRaw, p.db)
+	} else {
+		err = addCertDetails(cert.Spec.Domain, certDetailsRaw, p.db)
+	}
```
This is going to be nitpicky, but would it be slightly cleaner to have a `saveCertDetails` function where you pass in `isRenewal` and do the logic to add vs update there?

aalvarado1 replied:
Yeah I guess we can also do it that way. The only thing is that it will make the saveCertDetails func a bit bigger and I wanted to keep them simple.

aalvarado1 replied:
It would look something like 
`func saveCertDetails(domain string, certDetailsRaw []byte, d *gorm.DB, renewal bool) (err error) {
	s := string(certDetailsRaw)
	if renewal {
		var certDetails CertDetail
		if dbErr := d.Where(&CertDetail{Domain: domain}).First(&certDetails); dbErr != nil {
			return errors.Wrapf(err, "GORM error %s", dbErr)
		}
		certDetails.Value = s
		d.Save(&certDetails)
	} else {
		if dbErr := d.Create(&CertDetail{Domain: domain, Value: s}); dbErr != nil {
			return errors.Wrapf(err, "GORM error %s", dbErr)
		}
	}
	return err
}`

aalvarado1 replied:
Updated db.go and added a save function.

russellmacshane commented on processor.go:
```diff
@@ -624,15 +623,26 @@ func (p *CertProcessor) processCertificate(cert Certificate, forMaint bool) (boo
+	if isRenewal {
+		err = updateAltNames(cert.Spec.Domain, altNamesRaw, p.db)
+	} else {
+		err = addAltNames(cert.Spec.Domain, altNamesRaw, p.db)
```
Same suggestion as CertDetails above.

aalvarado1 replied:
Updated db.go and added a save function.

Commit:
Changes from review.
1. Removed logging in db.go for standard operations
2. Updated error handling to catch gorm errors
3. Moved and added defer db.Close() right after creation in main
4. Refactor of db connection args into separate vars
5. If db connection not possible app will panic

russellmacshane commented:
Need to add gorm to the glide yaml. This should do the trick. `glide get github.com/jinzhu/gorm`

Commit:
If record was not found it would throw and error and cert-manager
would not provision cert. Added check for record not found. Also some
tweaks and clean up from review.

ctriv commented on db.go:
```diff
@@ -0,0 +1,171 @@
+		return errors.Wrapf(dbErr, "Unable to update alt names for %s", domain)
+	}
+	altNames.Value = s
+	d.Save(&altNames)
```
You need to check the Save call for errors as well.

Welcome to go :trollface:.

aalvarado1 replied:
doh!

ctriv commented on db.go:
```diff
@@ -0,0 +1,171 @@
+		return errors.Wrapf(dbErr, "Unable to update cert details for %s", domain)
+	}
+	certDetails.Value = s
+	d.Save(&certDetails)
```
Error checking here too.

aalvarado1 replied:
double doh!

russellmacshane commented on db.go:
```diff
@@ -0,0 +1,171 @@
+/**
+Wrapper function to either save or update depending on renewal flag
+*/
+func saveCertDetails(domain string, certDetailsRaw []byte, d *gorm.DB, isRenewal bool) error {
```
💋 Perfection

aalvarado1 replied:
Excellent

Commit:
Forgot to add error check on save.

ctriv reviewed (approved):
Looks good.
